# ML Lab Sheet - 5 📑

### Name: Anish Agarwal
### Batch: 3NC8
### Roll No: 101915182

## Importing Relevant Libraries

In [1]:
import numpy as np
import pandas as pd
import matplotlib.pyplot as plt

## 🔰 Question: 1

## Importing DataSet

In [2]:
data = pd.read_csv('https://raw.githubusercontent.com/anishaga/Machine-Learning/main/Lab%20Sheet%205/Wine.csv')
X = data.iloc[:,:-1].values
y = data.iloc[:,-1].values

## Splitting The DataSet Into Training And Testing

In [3]:
from sklearn.model_selection import train_test_split
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size = 0.20, random_state = 0)

## Feature Scaling

In [4]:
from sklearn.preprocessing import StandardScaler
sc = StandardScaler()
X_train = sc.fit_transform(X_train)
X_test = sc.transform(X_test)

## Applying PCA

In [5]:
from sklearn.decomposition import PCA
pca = PCA(n_components = 2)
X_train = pca.fit_transform(X_train)
X_test = pca.transform(X_test)

## Training Logistic Regression Model On The Testing Data Set

In [6]:
from sklearn.linear_model import LogisticRegression
classifier = LogisticRegression()
classifier.fit(X_train, y_train)

LogisticRegression()

## Predicting Test Set Values 

In [7]:
y_pred = classifier.predict(X_test)

## Making Confusion Matrix

In [8]:
from sklearn.metrics import confusion_matrix
cm = confusion_matrix(y_test, y_pred);
print(cm)

[[14  0  0]
 [ 1 15  0]
 [ 0  0  6]]


## Finding Accuracy Score

In [9]:
from sklearn.metrics import accuracy_score
accuracy_score(y_test, y_pred)

0.9722222222222222

## 🔰 Question: 2

## Importing DataSet

In [10]:
data = pd.read_csv('https://raw.githubusercontent.com/anishaga/Machine-Learning/main/Lab%20Sheet%205/Wine.csv')
X = data.iloc[:,:-1].values
y = data.iloc[:,-1].values

## Splitting The DataSet Into Training And Testing

In [11]:
from sklearn.model_selection import train_test_split
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size = 0.3, random_state = 2)

## Feature Scaling

In [12]:
from sklearn.preprocessing import StandardScaler
sc = StandardScaler()
X_train = sc.fit_transform(X_train)
X_test = sc.transform(X_test)

## Applying LDA

In [13]:
from sklearn.discriminant_analysis import LinearDiscriminantAnalysis as LDA
lda = LDA(n_components = 2)
X_train = lda.fit_transform(X_train, y_train)
X_test = lda.transform(X_test)

## Training Logistic Regression Model On The Testing Data Set¶

In [14]:
from sklearn.linear_model import LogisticRegression
classifier = LogisticRegression()
classifier.fit(X_train, y_train)

LogisticRegression()

## Predicting Test Set Values

In [15]:
y_pred = classifier.predict(X_test)

## Making Confusion Matrix

In [16]:
from sklearn.metrics import confusion_matrix
cm = confusion_matrix(y_test, y_pred);
print(cm)

[[20  0  0]
 [ 0 20  0]
 [ 0  1 13]]


## Finding Accuracy Score

In [17]:
from sklearn.metrics import accuracy_score
accuracy_score(y_test, y_pred)

0.9814814814814815

## 🔰 Question: 3

We took the same dataset - `Wine.csv` for both PCA And LDA models

- `PCA` had an accuracy of: `97.22%`
- `LDA` had an accuracy of: `98.14%`

- PCA focuses on capturing the direction of maximum variation in the data set.
- PCA performs better in case where number of samples per class is less. 

- LDA focuses on finding a feature subspace that maximizes the separability between the groups
- LDA works better with large dataset having multiple classes; class separability is an important factor while reducing dimensionality.